src/data/dataset.c

```
@@ -0,0 +1,138 @@
+#include <string.h>
+#include <stdlib.h>
+
+#ifdef CUDA_AVAILABLE
+#include "core/arena_cuda.h"
+#endif
+
+TensorDataset *create_tensor_dataset(Tensor **tensors, u64 num_tensors) {
+  if (num_tensors == 0 || tensors == NULL) return NULL;
+  
+  TensorDataset *dataset = malloc(sizeof(TensorDataset));
+  dataset->num_tensors = num_tensors;
```

Cppcheck (2.20.0)

[warning] If memory allocation fails, then there is a possible null pointer dereference

(nullPointerOutOfMemory)